### ISSUE TITLE
armv7l compile issues with arm_neon.h

### ISSUE THREAD

**Issue by l3eta** [kcat/openal-soft#54]
Make Output:
[alarm@alarm build]$ make
Scanning dependencies of target makehrtf
[  1%] Building C object CMakeFiles/makehrtf.dir/utils/makehrtf.c.o
[  3%] Linking C executable makehrtf
[  3%] Built target makehrtf
Scanning dependencies of target common
[  4%] Building C object CMakeFiles/common.dir/common/almalloc.c.o
[  6%] Building C object CMakeFiles/common.dir/common/atomic.c.o
[  8%] Building C object CMakeFiles/common.dir/common/rwlock.c.o
[  9%] Building C object CMakeFiles/common.dir/common/threads.c.o
[ 11%] Building C object CMakeFiles/common.dir/common/uintmap.c.o
[ 12%] Linking C static library libcommon.a
[ 12%] Built target common
Scanning dependencies of target openal
[ 14%] Building C object CMakeFiles/openal.dir/OpenAL32/alAuxEffectSlot.c.o
/media/sdcard/workspace/openal-soft/OpenAL32/alAuxEffectSlot.c: In function 'RemoveEffectSlotList':
/media/sdcard/workspace/openal-soft/OpenAL32/alAuxEffectSlot.c:431:9: warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
     if(!ATOMIC_COMPARE_EXCHANGE_STRONG(ALeffectslot*, &context->ActiveAuxSlotList, &root, next))
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/media/sdcard/workspace/openal-soft/OpenAL32/alAuxEffectSlot.c:437:18: warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
         } while(!ATOMIC_COMPARE_EXCHANGE_STRONG(ALeffectslot*, &cur->next, &root, next));
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 16%] Building C object CMakeFiles/openal.dir/OpenAL32/alBuffer.c.o
[ 17%] Building C object CMakeFiles/openal.dir/OpenAL32/alEffect.c.o
[ 19%] Building C object CMakeFiles/openal.dir/OpenAL32/alError.c.o
[ 20%] Building C object CMakeFiles/openal.dir/OpenAL32/alExtension.c.o
[ 22%] Building C object CMakeFiles/openal.dir/OpenAL32/alFilter.c.o
[ 24%] Building C object CMakeFiles/openal.dir/OpenAL32/alListener.c.o
[ 25%] Building C object CMakeFiles/openal.dir/OpenAL32/alSource.c.o
[ 27%] Building C object CMakeFiles/openal.dir/OpenAL32/alState.c.o
[ 29%] Building C object CMakeFiles/openal.dir/OpenAL32/alThunk.c.o
[ 30%] Building C object CMakeFiles/openal.dir/OpenAL32/sample_cvt.c.o
[ 32%] Building C object CMakeFiles/openal.dir/Alc/ALc.c.o
[ 33%] Building C object CMakeFiles/openal.dir/Alc/ALu.c.o
[ 35%] Building C object CMakeFiles/openal.dir/Alc/alcConfig.c.o
[ 37%] Building C object CMakeFiles/openal.dir/Alc/alcRing.c.o
[ 38%] Building C object CMakeFiles/openal.dir/Alc/bs2b.c.o
[ 40%] Building C object CMakeFiles/openal.dir/Alc/effects/chorus.c.o
[ 41%] Building C object CMakeFiles/openal.dir/Alc/effects/compressor.c.o
[ 43%] Building C object CMakeFiles/openal.dir/Alc/effects/dedicated.c.o
[ 45%] Building C object CMakeFiles/openal.dir/Alc/effects/distortion.c.o
[ 46%] Building C object CMakeFiles/openal.dir/Alc/effects/echo.c.o
[ 48%] Building C object CMakeFiles/openal.dir/Alc/effects/equalizer.c.o
[ 50%] Building C object CMakeFiles/openal.dir/Alc/effects/flanger.c.o
[ 51%] Building C object CMakeFiles/openal.dir/Alc/effects/modulator.c.o
[ 53%] Building C object CMakeFiles/openal.dir/Alc/effects/null.c.o
[ 54%] Building C object CMakeFiles/openal.dir/Alc/effects/reverb.c.o
[ 56%] Building C object CMakeFiles/openal.dir/Alc/helpers.c.o
[ 58%] Building C object CMakeFiles/openal.dir/Alc/bsinc.c.o
[ 59%] Building C object CMakeFiles/openal.dir/Alc/hrtf.c.o
[ 61%] Building C object CMakeFiles/openal.dir/Alc/uhjfilter.c.o
[ 62%] Building C object CMakeFiles/openal.dir/Alc/ambdec.c.o
[ 64%] Building C object CMakeFiles/openal.dir/Alc/bformatdec.c.o
[ 66%] Building C object CMakeFiles/openal.dir/Alc/panning.c.o
[ 67%] Building C object CMakeFiles/openal.dir/Alc/mixer.c.o
[ 69%] Building C object CMakeFiles/openal.dir/Alc/mixer_c.c.o
[ 70%] Building C object CMakeFiles/openal.dir/Alc/mixer_neon.c.o
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c: In function 'ApplyCoeffsStep':
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:6169:1: error: inlining failed in call to always_inline 'vcombine_f32': target specific option mismatch
 vcombine_f32 (float32x2_t __a, float32x2_t __b)
 ^~~~~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:24:20: note: called from here
         leftright4 = vcombine_f32(leftright2, leftright2);
         ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:5487:1: error: inlining failed in call to always_inline 'vset_lane_f32': target specific option mismatch
 vset_lane_f32 (float32_t __a, float32x2_t __b, const int __c)
 ^~~~~~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:23:20: note: called from here
         leftright2 = vset_lane_f32(right, leftright2, 1);
         ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:5487:1: error: inlining failed in call to always_inline 'vset_lane_f32': target specific option mismatch
 vset_lane_f32 (float32_t __a, float32x2_t __b, const int __c)
 ^~~~~~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:22:20: note: called from here
         leftright2 = vset_lane_f32(left, leftright2, 0);
         ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:5714:1: error: inlining failed in call to always_inline 'vdup_n_f32': target specific option mismatch
 vdup_n_f32 (float32_t __a)
 ^~~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:21:21: note: called from here
         float32x2_t leftright2 = vdup_n_f32(0.0);
                     ^~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:9517:1: error: inlining failed in call to always_inline 'vst1q_f32': target specific option mismatch
 vst1q_f32 (float32_t * __a, float32x4_t __b)
 ^~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:40:9: note: called from here
         vst1q_f32(&Coeffs[c][0], coefs);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:9434:1: error: inlining failed in call to always_inline 'vst1_f32': target specific option mismatch
 vst1_f32 (float32_t * __a, float32x2_t __b)
 ^~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:39:9: note: called from here
         vst1_f32((float32_t*)&Values[o1][0], vget_high_f32(vals));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:6252:1: error: inlining failed in call to always_inline 'vget_high_f32': target specific option mismatch
 vget_high_f32 (float32x4_t __a)
 ^~~~~~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:39:9: note: called from here
         vst1_f32((float32_t*)&Values[o1][0], vget_high_f32(vals));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:9434:1: error: inlining failed in call to always_inline 'vst1_f32': target specific option mismatch
 vst1_f32 (float32_t * __a, float32x2_t __b)
 ^~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:38:9: note: called from here
         vst1_f32((float32_t*)&Values[o0][0], vget_low_f32(vals));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:6320:1: error: inlining failed in call to always_inline 'vget_low_f32': target specific option mismatch
 vget_low_f32 (float32x4_t __a)
 ^~~~~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:38:9: note: called from here
         vst1_f32((float32_t*)&Values[o0][0], vget_low_f32(vals));
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:595:1: error: inlining failed in call to always_inline 'vaddq_f32': target specific option mismatch
 vaddq_f32 (float32x4_t __a, float32x4_t __b)
 ^~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:36:15: note: called from here
         coefs = vaddq_f32(coefs, deltas);
         ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:1330:1: error: inlining failed in call to always_inline 'vmlaq_f32': target specific option mismatch
 vmlaq_f32 (float32x4_t __a, float32x4_t __b, float32x4_t __c)
 ^~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:35:14: note: called from here
         vals = vmlaq_f32(vals, coefs, leftright4);
         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:9017:1: error: inlining failed in call to always_inline 'vld1q_f32': target specific option mismatch
 vld1q_f32 (const float32_t * __a)
 ^~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:33:21: note: called from here
         float32x4_t deltas = vld1q_f32(&CoeffStep[c][0]);
                     ^~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:9017:1: error: inlining failed in call to always_inline 'vld1q_f32': target specific option mismatch
 vld1q_f32 (const float32_t * __a)
 ^~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:32:21: note: called from here
         float32x4_t coefs = vld1q_f32((float32_t*)&Coeffs[c][0]);
                     ^~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:6169:1: error: inlining failed in call to always_inline 'vcombine_f32': target specific option mismatch
 vcombine_f32 (float32x2_t __a, float32x2_t __b)
 ^~~~~~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:30:21: note: called from here
         float32x4_t vals = vcombine_f32(vld1_f32((float32_t*)&Values[o0][0]),
                     ^~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:8934:1: error: inlining failed in call to always_inline 'vld1_f32': target specific option mismatch
 vld1_f32 (const float32_t * __a)
 ^~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:30:28: note: called from here
         float32x4_t vals = vcombine_f32(vld1_f32((float32_t*)&Values[o0][0]),
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                         vld1_f32((float32_t*)&Values[o1][0]));
                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:3:0:
/usr/lib/gcc/armv7l-unknown-linux-gnueabihf/6.1.1/include/arm_neon.h:8934:1: error: inlining failed in call to always_inline 'vld1_f32': target specific option mismatch
 vld1_f32 (const float32_t * __a)
 ^~~~~~~~
/media/sdcard/workspace/openal-soft/Alc/mixer_neon.c:30:28: note: called from here
         float32x4_t vals = vcombine_f32(vld1_f32((float32_t*)&Values[o0][0]),
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                         vld1_f32((float32_t*)&Values[o1][0]));
                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/openal.dir/build.make:903: CMakeFiles/openal.dir/Alc/mixer_neon.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:105: CMakeFiles/openal.dir/all] Error 2
make: *** [Makefile:128: all] Error 2




```
#Check for ARM Neon support
OPTION(ALSOFT_REQUIRE_NEON "Require ARM Neon support" OFF)
CHECK_INCLUDE_FILE(arm_neon.h HAVE_ARM_NEON_H)
IF(HAVE_ARM_NEON_H)
    OPTION(ALSOFT_CPUEXT_NEON "Enable ARM Neon support" ON)
    IF(ALSOFT_CPUEXT_NEON)
        SET(HAVE_NEON 1)
SET(ALC_OBJS  ${ALC_OBJS} Alc/mixer_neon.c)
SET(CPU_EXTS "${CPU_EXTS}, Neon")
ENDIF()
ENDIF()
IF(ALSOFT_REQUIRE_NEON AND NOT HAVE_NEON)
    MESSAGE(FATAL_ERROR "Failed to enabled required ARM Neon CPU extensions")
ENDIF()
```

Commenting the above out fixes it enough to allow me to build and use the .so in lwjgl 2.9.4

**Comment by kcat** [comment]
Looks like the target being compiled for doesn't support those Neon functions. I'm not sure what the default target is like, but I often see people using ```-mfloat-abi=softfp -mfpu=neon``` for Neon to work. I'm not too familiar with building for ARM/Android though, so I don't know the implications of such compile options.

FWIW, you can also pass ```-DALSOFT_CPUEXT_NEON=OFF``` to cmake to disable the use of Neon extensions (though it will use less efficient mixer routines than the Neon-based ones).

**Comment by l3eta** [comment]
How would I go about getting neon to compile then?

**Comment by kcat** [comment]
Try ```cmake .. -DCMAKE_C_FLAGS="-mfpu=neon"```. You may need to delete ```CMakeCache.txt``` first.

**Comment by AndreRH** [comment]
This also happens in buildroot ...
The underlying problem is the quite simple check for the header arm_neon.h which is present in most toolchains and is a bad indicator for a neon build, rather a compilation check should be used i think or some other check to see if the target fpu is neon...

**Comment by arnout** [comment]
Actually, the problem is that the default CFLAGS may contain e.g. -mfpu=vfpv3, to tell the compiler that it can use VFP instructions. However, when mixing VFP instructions with NEON instructions, the registers must be saved. Therefore the compiler doesn't allow mixing them within the same function.

The always_inline attribute of the NEON intrinsics will cause this failure when the surrounding function is built with -mfpu=vfpv3.

The solution is rather simple: explicitly pass -mfpu=neon for the files that contain NEON intrinsics. Or perhaps for all of openal, assuming that it's OK for the floating point arithmetic to violate IEEE 754. Better test that :-)

Note that there is no easy way to be sure that the target processor really supports NEON.  When you're building for e.g. Cortex-A5, the compiler will always accept -mfpu=neon, because a Cortex-A5 may or may not have NEON support. So it's best to only turn on the NEON option when explicitly configured to do so with ALSOFT_CPUEXT_NEON. If not, you need to add a runtime detection mechanism.

**Comment by kcat** [comment]
One option then would be to force ```-mfpu=neon``` on sources that explicitly use NEON, similar to how ```-msse``` is forced on sources that explicitly use SSE. Would that have any negative consequences, for instance if the FPU is already set to something "better"? Would it create ABI problems if some functions are built with `-mfpu=neon```, and other functions are built without it and call those NEON functions?

What's the proper way to do runtime detection for NEON? I don't suppose some form of CPUID is available...

**Comment by tpetazzoni** [comment]
I've submitted https://github.com/kcat/openal-soft/pull/64 to address this issue.

**Comment by sneakernets** [comment]
Thanks for tracking this down, @tpetazzoni .

**Comment by kcat** [comment]
Should be fixed now with commit 4d91afb20302536c068d34c3c54a5380c2f750f2. There is also commit c3c283a0b5d0130afafaa2a5b6ce6fbc30b6e6a1 which attempts to do run-time detection for NEON extensions, to avoid it crashing on systems without NEON but with libs that have support for NEON-enhanced routines.

**Comment by yangjiel** [comment]
Could you backport this to version 1.15.1? Since Openembedded is using 1.15.1 on morty branch, the branch is close, I think I cannot change the version of the open-soft but I can make a patch just change the check sum to fetch the newest 1.15.1 to get it to work with armv7.
